File: BT4.cpp
#include <iostream>
#include <stack>
using namespace std;
int main () {
    int n, q;cin >> n>> q;
    int a[n + 1];
    for (int i = 1; i <= n; i++) cin >> a[i];
    int r[n  + 5];
    stack<int> st;
    for (int i = n; i >= 1; i --) {
        while(not st.empty() and a[i] >= st.top()) {
            st.pop();
        }
        if (st.empty()){
            r[i] = 0;
        }
        else{
            r[i] = st.size();
        }
        st.push(a[i]);
    }

    for(int i = 0; i < q; i++) {
        int x; cin >> x;
        cout << r[x] << endl;
    }
}

File: BT4.py
from collections import deque
def function():
    n, q = map(int,input().split())
    ls = [0]
    while len(ls) < n + 1:
        ls += list(map(int, input().split())) + [-1]
    stack = deque([0])
    dp = [0] * (n + 5)

    for i in range(n, 0, - 1):
        while stack and ls[i] >= stack[-1]:
            stack.pop()
        if not stack:
            dp[i] = 0
        else:
            dp[i] = len(stack)
        stack.append(ls[i])
    for i in range(q):
        print(dp[int(input())])
if __name__ == "__main__":
    function()